tokio/src/sync/tests/loom_oneshot.rs

```diff
@@ -138,3 +140,48 @@ fn changing_tx_task() {
         }
     });
 }
+
+#[test]
+fn checking_tx_send_ok_not_drop() {
+    use std::cell::RefCell;
+    loom::thread_local! {
+        static IS_RX: RefCell<bool> = RefCell::new(true);
+    }
+
+    struct Msg;
+
+    impl Drop for Msg {
+        fn drop(&mut self) {
+            IS_RX.with(|is_rx: &RefCell<_>| {
+                // On `tx.send(msg)` returning `Err(msg)`,
+                // we call `std::mem::forget(msg)`, so that
+                // `drop` is not expected to be called in the
+                // tx thread.
+                assert!(*is_rx.borrow());
+            });
+        }
+    }
+
+    loom::model(|| {
+        let (tx, rx) = oneshot::channel();
+
+        // tx thread
+        let tx_thread_join_handle = thread::spawn(move || {
+            // Ensure that `Msg::drop` in this thread will see is_rx == false
+            IS_RX.with(|is_rx: &RefCell<_>| {
+                *is_rx.borrow_mut() = false;
+            });
+            if let Err(msg) = tx.send(Msg) {
+                std::mem::forget(msg);
+            }
+        });
+
+        // rx thread
+        let rx_thread_join_handle = thread::spawn(move || {
+            drop(rx);
+        });
```

Comment:
You can just call `drop(rx)` on the main thread. Spawning threads is very expensive in loom tests.

Reply:
I had a try. 

If I simply use `drop(rx)` in the main thread, when I revert the changes in `oneshot.rs`, the assertion cannot be hit, and if I do it in a spawned thread, the assertion can be hit. 

I doubt that only `drop(rx)` in a separate thread can cover the case in loom.

Reply:
Do you hit the assertion with `LOOM_MAX_PREEMPTIONS=2`? That's what we run with in CI.

Reply:
Oops, when I set it to 2, the assertion can be hit.

Should the `CONTRIBUTING.md` be updated? I ran the loom test following the guide in it. 

https://github.com/tokio-rs/tokio/blob/df7706348554a039946925f7492d98e31bd46a85/CONTRIBUTING.md?plain=1#L195

Reply:
It takes a lot longer to run with the option set to 2, and most bugs can be caught with the value set to 1. It may be better to have your test use [the loom test builder][1] to explicitly set the max preemptions to 2.

[1]: https://docs.rs/loom/latest/loom/model/struct.Builder.html

Reply:
LGTM. Have updated the code.